code/datums/elements/footstep.dm

```diff
@@ -0,0 +1,152 @@
+///Footstep element. Plays footsteps at parents location when it is appropriate.
+/datum/element/footstep
+	element_flags = ELEMENT_DETACH|ELEMENT_BESPOKE
+	id_arg_index = 2
+	///A list containing living mobs and the number of steps they have taken since the last time their footsteps were played.
+	var/list/steps_for_living = list()
+	///volume determines the extra volume of the footstep. This is multiplied by the base volume, should there be one.
+	var/volume
+	///e_range stands for extra range - aka how far the sound can be heard. This is added to the base value and ignored if there isn't a base value.
+	var/e_range
+	///footstep_type is a define which determines what kind of sounds should get chosen.
+	var/footstep_type
+	///This can be a list OR a soundfile OR null. Determines whatever sound gets played.
+	var/footstep_sounds
+	///Whether or not to add variation to the sounds played
+	var/sound_vary = FALSE
+
+/datum/element/footstep/Attach(datum/target, footstep_type = FOOTSTEP_MOB_BAREFOOT, volume = 0.5, e_range = -8, sound_vary = FALSE)
+	. = ..()
+	if(!ismovable(target))
+		return ELEMENT_INCOMPATIBLE
+	src.volume = volume
+	src.e_range = e_range
+	src.footstep_type = footstep_type
+	src.sound_vary = sound_vary
+	switch(footstep_type)
+		if(FOOTSTEP_MOB_HUMAN)
+			if(!ishuman(target))
+				return ELEMENT_INCOMPATIBLE
+			RegisterSignal(target, COMSIG_MOVABLE_MOVED, PROC_REF(play_humanstep))
+			steps_for_living[target] = 0
+			return
+		if(FOOTSTEP_MOB_CLAW)
+			footstep_sounds = GLOB.clawfootstep
+		if(FOOTSTEP_MOB_BAREFOOT)
+			footstep_sounds = GLOB.barefootstep
```

Comment:
Instead of making large switch statements like this, why not just define subtypes with different vars or a proc override, that way this is not necessary. You can just directly give a typepath for the footstep element then.

Reply:
example?

Reply:
make a subtype of /datum/element/footstep for each type, and then just override proc behavior or change a var on the parent type

Reply:
So, subtyping like this?

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/62388554/4727bac6-6380-40f7-b352-269a09d1f16c)

Im still somewhat confused as to what the benefit this is though.

Would I change the element being called in each different mob's initialize to the specific subtype? That seems more difficult than just changing a single var for the specific mob like so 

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/62388554/283bd127-f7be-4a73-bf49-fafe4a1e0cac)

I just feel like Im inflating both the element, and making its calls more complex

Am I missing something? Tell me your thought process

Reply:
You can just change `footstep_type` to hold a typepath to the new element subtype. `footstep-type = /datum/element/claw`, and then during init for all mobs do `if(footstep_type) AttachElement(footstep_type)`

Reply:
Yeah, I dont see how this is any better than the switch. Ive tried subtyping the element as suggested, and it just bloats the amount of duplicate code more than it needs to be, especially when overriding the attach proc.